File: 0980-unique-paths-iii/0980-unique-paths-iii.cpp
class Solution {
private:
    int dx[4] = {0, 1, 0, -1};
    int dy[4] = {1, 0, -1, 0};

    bool isValid(int x, int y, vector<vector<int>>& grid){
            return x>=0 && y>=0 && x<grid.size() && y<grid[0].size();
    }
    void backtrack(int x, int y, int& target, int tx, int ty, vector<vector<int>>& grid, vector<vector<bool>>& visited, int& ans){
        if(x==tx && y==ty && target==-1){
            ans+=1;
            return;
        }
        
        for(int i=0;i<4;i++){
            int newx = x+dx[i];
            int newy = y+dy[i];
            
            if(isValid(newx, newy, grid) && !visited[newx][newy] && grid[newx][newy]!=-1){
                visited[newx][newy]=true;
                target-=1;
                backtrack(newx, newy, target, tx, ty, grid, visited, ans);
                target+=1;
                visited[newx][newy]=false;
            }
        }
    }
public:
    int uniquePathsIII(vector<vector<int>>& grid) {
        int target=0;
        int sx, sy, tx, ty;
        int m =grid.size(), n = grid[0].size();
        vector<vector<bool>> visited(m, vector<bool>(n, false));
        
        for(int i=0;i<m;i++){
            for(int j=0;j<n;j++){
                if(grid[i][j]==1){
                    sx=i;
                    sy=j;
                } else if(grid[i][j]==2){
                    tx=i;
                    ty=j;
                } else if(grid[i][j]==0){
                    target+=1;
                }
            }
        }
        visited[sx][sy]=true;
        int ans=0;
        backtrack(sx, sy, target, tx, ty, grid, visited, ans);
        return ans;
    }
};

File: 0980-unique-paths-iii/0980-unique-paths-iii.ts
const dx: number[] = [0, 1, 0, -1];
const dy: number[] = [1, 0, -1, 0];

function isValid(x:number, y:number, grid:number[][]): boolean{
    return x>=0 && y>=0 && x<grid.length && y<grid[0].length;
}

function dfs(x: number, y: number, tx: number, ty: number, grid:number[][], squares:number, target:number) : number {
    if(x==tx && y==ty && squares == target+1) return 1;
    
    let ans=0;
    for(let k=0;k<4;k++){
        let new_x = x + dx[k];
        let new_y = y + dy[k];
        
        if(isValid(new_x, new_y, grid) && grid[new_x][new_y]!=-1){
            let temp = grid[new_x][new_y]; 
            grid[new_x][new_y] = -1;
            squares+=1;
            ans += dfs(new_x, new_y, tx, ty, grid, squares, target);
            squares-=1;
            grid[new_x][new_y] = temp;   
        }  
    }
    return ans;
}

function uniquePathsIII(grid: number[][]): number {
    let target=0, squares=0;
    let sx, sy, tx, ty;
    
    for(let i=0;i<grid.length;i++){
        for(let j=0;j<grid[0].length;j++){
            if(grid[i][j] == 1){
                [sx, sy] = [i, j];
            } else if(grid[i][j] == 2){
                [tx, ty] = [i, j];
            } else if(grid[i][j] == 0){
                target+=1;
            }
        }
    }
    
    grid[sx][sy]=-1;
    return dfs(sx, sy, tx, ty, grid, squares, target);
    
};